fuel-vm/src/tests/blockchain.rs

```diff
@@ -607,26 +607,170 @@ fn ldc_contract_not_in_inputs() {
+
+    let program_ops = vec![
+        op::movi(0x10, 0x11),
+        op::movi(0x11, 0x2a),
+        op::add(0x12, 0x10, 0x11),
+        op::log(0x10, 0x11, 0x12, 0x00),
+        op::ret(0x20),
     ];
 
-    ldc_reason_helper(load_contract, MemoryOverflow, true);
+    let program = program_ops.clone().into_iter().collect::<Vec<u8>>();
+    let contract_size = program.len();
+    let contract_id = test_context
+        .setup_contract(program_ops, None, None)
+        .contract_id;
+
+    let (script, _) = script_with_data_offset!(
+        data_offset,
+        vec![
+            op::movi(0x10, 2048),
+            op::aloc(0x10),
+            op::addi(0x10, RegId::HP, contract_size as Immediate12),
+            op::movi(0x10, 1234),
+            op::addi(0x10, RegId::HP, (contract_size + 1) as Immediate12),
+            op::movi(0x10, 1234),
+            op::movi(0x20, data_offset as Immediate18),
+            op::add(0x11, RegId::ZERO, 0x20),
+            op::movi(0x13, (contract_size + 2) as Immediate18),
+            op::ccp(RegId::HP, 0x11, RegId::ZERO, 0x13),
+            op::addi(0x21, 0x20, ContractId::LEN as Immediate12),
+            op::meq(0x30, 0x21, RegId::HP, 0x13),
+            op::ret(0x30),
+        ],
+        TxParameters::DEFAULT.tx_offset()
+    );
+
+    let mut script_data = contract_id.to_vec();
+    script_data.extend(program.as_slice());
+    script_data.extend(vec![0; 2]);
+
+    let result = test_context
+        .start_script(script, script_data)
+        .gas_limit(gas_limit)
+        .contract_input(contract_id)
+        .fee_input()
+        .contract_output(&contract_id)
+        .execute();
+
+    let receipts = result.receipts();
+    let ret = receipts
+        .first()
+        .expect("A `RET` opcode was part of the program.");
+
+    assert_eq!(1, ret.val().expect("A constant `1` was returned."));
+}
+
+#[test]
+fn code_copy_out_of_contract_offset_over_length() {
+    let mut test_context = TestBuilder::new(2322u64);
+    let gas_limit = 1_000_000;
+
+    let program_ops = vec![
+        op::movi(0x10, 0x11),
+        op::movi(0x11, 0x2a),
+        op::add(0x12, 0x10, 0x11),
+        op::log(0x10, 0x11, 0x12, 0x00),
+        op::ret(0x20),
+    ];
+
+    let program = program_ops.clone().into_iter().collect::<Vec<u8>>();
+    let contract_size = program.len();
+    let contract_id = test_context
+        .setup_contract(program_ops, None, None)
+        .contract_id;
+
+    let (script, _) = script_with_data_offset!(
+        data_offset,
+        vec![
+            op::movi(0x10, 2048),
+            op::aloc(0x10),
+            op::movi(0x20, data_offset as Immediate18),
+            op::add(0x11, RegId::ZERO, 0x20),
+            op::movi(0x12, (contract_size + 1) as Immediate18),
+            op::movi(0x13, contract_size as Immediate18),
+            op::ccp(RegId::HP, 0x11, 0x12, 0x13),
+            op::addi(0x21, 0x20, ContractId::LEN as Immediate12),
+            op::meq(0x30, 0x21, RegId::HP, 0x13),
```

Comment:
Instead of checking if the result of ccp is != contract bytecode, maybe we should have a positive check that the ccp result zero filled as expected? Otherwise this test could still pass even if ccp just returned a bunch of garbage when the offset is > contract length.

Reply:
Good catch. Fixed in 0c3bb19.